rust/tool/src/install.rs

```diff
@@ -243,6 +317,61 @@ impl Installer {
     }
 }
 
+/// Stores the source and destination of all artifacts needed to install all generations.
+///
+/// This enables a two step installaton process where all artifacts across all generations are
+/// first collected and then installed. This allows deduplication in the collection phase and thus
+/// reduces the number of accesesses and writes to the ESP.
+///
+/// The underlying data structure to store the paths is a HashMap to map source and destination to
+/// each other while automatically ensuring uniqueness.
+struct GenerationArtifacts {
+    /// Temporary directory that stores all temporary files that are created when building the
+    /// GenerationArtifacts.
+    tempdir: TempDir,
+    /// Mapping of signed files from their destinations to their source.
+    signed_files: HashMap<PathBuf, PathBuf>,
+    /// Mapping of unsigned files from their destinations to their source.
+    unsigned_files: HashMap<PathBuf, PathBuf>,
+}
```

Comment:
There a couple of issues here.

- On the more tactical level, the data structure seems awkward. Why not have a `files_to_copy: Vec<(PathBuf, PathBuf)>` and a `files_to_sign: Set<PathBuf>` or one `Vec` of `struct FileAction { source, destination, needs_signing }`.
- It makes the code hard to understand if a mutable reference of `GenerationArtifacts` is handed around. I think this happens, because `tmpdir` needs to be there to already write `os-release` files? If so, `source` in `FileAction` could be an existing file or a string that needs to be written out at the destination.
- On a higher-level, it seems we are mixing up the "trying to figure out what to do" part, which could be completely functional without mutable state, with the "let's do it" part. Is it possible to refactor the code so we can build up what we want in memory first and then just hand this to the "execute" function? This may make everything easier to test as well.

Reply:
- The key feature of the HashMap is that it automatically deduplicates the unsigned files based on their destination path (i.e. the path on the ESP). Thus, when we are building the signed files, all unsigned files on the ESP are already uniquely determined. This way, we can generate UKIs while being sure that the hash embedded in them will point to a valid file on the ESP because the file will not be overwritten by a later generation. None of the data structures you suggested have this characteristic. We could use something like `HashMap<PathBuf, FileAction>` with `struct FileAction { source, needs_signing }` and then move the `tempdir` into `Installer`. The tempdir would then live longer than necessary, but I think that's fine.
- The `tempdir` is actually immutable as it is just a path. The entire struct is mutable because unsigned_files and signed_files are modified. The tempdir is really only part of this struct to not have to pass it to every function separately. If we implement `GenerationArtifacts(HashMap<PathBuf, FileAction>)` then it would still need to be mutable so that the build_* functions can insert into the HashMap.
- We can certainly separate the planning from the executing. However, this would be a much larger change than this already is and it would require a more complex data structure for the plan. For example, see [bootspec-secureboot](https://github.com/DeterminateSystems/bootspec-secureboot/blob/main/installer/src/systemd_boot/plan.rs). I see this change as a first step to separate planning from execution.

Reply:
> The key feature of the HashMap is that it automatically deduplicates the unsigned files

This is a good point. This should be more explicit in the code. If you can drop this high-level "flow" as documentation somewhere, this would really help the reader.

> We can certainly separate the planning from the executing. However, this would be a much larger change than this already is and it would require a more complex data structure for the plan.

Yes. After sleeping about it, I also realized that this request just blocks this much needed improvement for the sake of perfection. I will re-read the [Google Code Review](https://google.github.io/eng-practices/review/reviewer/standard.html) guidelines. :)

Reply:
I hope I have improved the comments a bit now.